### PR TITLE
[hail][lsm] Teach the LSM how to create roughly even key partitions

### PR DESCRIPTION
**Pull request by danking**
This adds `partitionKeys` which partitions the keys into partitions containing as-near-to-equal-as-possible number of records.

The partitions are represented by an array, `pb`, of keys of length `nPartitions + 1`. The partitions are taken to include records with keys in
```
[pb(0), pb(1))
[pb(1), pb(2))
...
[pb(nPartitions-1), pb(nPartitions)]
```
Note that the last partition is inclusive of both end-points. I have two simple examples of the partitioning behavior at the top of `LSMSuite`. In these cases, the number of elements is so small that the LSM perfectly stored the distribution, so there is no approximation.

`LSMSuite` also contains tests that use enough keys so as to force the LSM to not keep them all. When the LSM has more than 10,000 keys, it starts sampling. It flips a coin that is true with probability `10,000 / n_keys_seen`. As we see more keys, the probability that the next seen key is kept decreases. If we decide to keep a key, we uniformly randomly choose a key to evict.

The above is not entirely true. In reality, we keep 9998 keys in an array and separately keep the greatest and the least. Those are the true greatest seen key and the true least seen key. The probabilities above are adjusted accordingly.

If the sample of keys is unbiased, then we expect the partitions chosen to be roughly equal in number of records.

---

The TestNG changes are already in another PR, I'll rebase when that lands.

I separately fixed a bug in KeyedCodecSpec wherein it incorrectly assumed the Key and Value types were the same.

I also fixed a bug in that the LSM used non-thread-local regions. Regions are not thread safe and the Indeed LSM implementation uses many threads.

In order to track every addition to the LSM, I made the Indeed LSM object private and made the Hail LSM class have the necessary methods.

### CODE REVIEW

**Comment by danking** [comment]
I promise the tests all work. It's this TestNG nonsense. I'll back out my use of DataProvider, which is apparently breaking TestNG 6.8.21, but any newer version of TestNG breaks our test jar.

File: hail/src/main/scala/is/hail/shuffler/LSM.scala [hunk 1]
```diff
@@ -122,8 +139,128 @@ class LSM (
...
+  private[this] var samplesEnd = 0
+
+  private[this] def maybeSample(k: Long): Unit = {
+    if      (processed == 0)
+      assert(least == -1 && greatest == -1)
+    else if (processed == 1)
+      assert(least != -1 && greatest == -1)
+    else if (processed >= 2)
+      assert(keyOrd.compare(least, greatest) < 0)
+    assert(samplesEnd <= samples.length)
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/shuffler/LSM.scala:161]: Are these assertions redundant based on the ones later in the function?

Comment (chrisvittal) [reply hail/src/main/scala/is/hail/shuffler/LSM.scala:161]: Either this, or are the later assertions redundant?

Comment (danking) [reply hail/src/main/scala/is/hail/shuffler/LSM.scala:161]: Sure are, I've removed the later ones.

**Pull request merged**